### DISCUSSION TITLE
RFC: Integrating project links into the Classifier component

### DISCUSSION THREAD

**Discussion by eatyourgreens** [zooniverse/front-end-monorepo#2165] [General]
For Engaging Crowds, we want to block the task area when you load a subject that' you've already classified, or which has been retired. Instead of classifying the subject, you'll be given the option to return to the project home page, navigate to the next (or previous) subject, or choose a new subject to view (or classify.) See https://github.com/zooniverse/front-end-monorepo/issues/2027 and the designs in the comments.

The classifier could import `next/link`, then write each of the links in that popup as something like:
```jsx
<Link href="/owner/project/" label="Project Home" />
```
That would make the Classifier component implicitly part of the project app, so it couldn't really be published and used as a self-contained component.

The router could be passed to the Classifier as a prop, like the client is now. Routes could then be handled by the store but this still ties the Classifier component to the specific routing of the NextJS project app.

I'm looking for a way to embed that popup in the Classifier component, without explicitly adding the project app routes to the Classifier. I wondered if anyone has any suggestions?

**Comment by srallen** [comment]
Going home could be done by: `location.assign(location.origin)` yeah? It'd be a full page refresh, but I think that's ok.

**Reply by eatyourgreens** [reply]
The problem with the full page refresh is you run `getServerSideProps()` again, so we pay for another build of the page and its CSS. Maybe if we did that but used `getStaticProps()` so we aren't hitting the API too much, that might be better.

I guess there's still the overhead of authenticating again, which you wouldn't have with a client-side transition. I haven't thought too much about it, but there might be other client-side state that's lost when the store is rebuilt.

**Reply by eatyourgreens** [reply]
Thinking about it a bit more, `location.assign(location.origin)` would send you to the Zooniverse home page, but we want you to go to the project home page eg. https://www.zooniverse.org/projects/msalmon/hms-nhs-the-nautical-health-service

**Reply by srallen** [reply]
We've discussed stashing snapshots of the store into session storage / local storage / indexedDB and rehydrating the store as needed for future features, primarily when navigating between Talk and the classify page, so I think that'd be a viable solution here too. We could think about doing this in memory server side too? Not sure what the mechanisms would be for that exactly.

**Reply by srallen** [reply]
Yeah, we can reconstruct that since the Project resource is stored which includes its slug, ```location.assign(`${location.origin}/projects/${project.slug}`)```, maybe use this if there's a boolean enabled, `useProjectSlug` as proxy for the classifier being within project app context? Or a store view which returns `/projects/${project.slug}` depending on configuration? Just a rough idea. I think it's worth keeping the classifier decoupled as much as possible here.

**Reply by eatyourgreens** [reply]
`getStaticProps()` generates a JSON snapshot of the store, which is saved as a file for however long we set the revalidation time to. We could definitely take advantage of that. It's passed to the browser as `initialState` and used to rehydrate the store here: https://github.com/zooniverse/front-end-monorepo/blob/582e6e48181d814f529d2c34e6969641cb5cbbeb/packages/app-project/pages/_app.js#L36-L37

**Comment by eatyourgreens** [comment]
Another possible pain point is the 'Subject Selection Modal' link, which relates to https://github.com/zooniverse/front-end-monorepo/discussions/2166

![Screenshot of the task area modal, showing links for Project Home and Subject Selection and buttons for loading a new subject.](https://user-images.githubusercontent.com/1392130/107423709-48b66c00-6ae2-11eb-98ca-a403506855af.png)

You can open subject set selection by linking directly to a workflow eg. https://www.zooniverse.org/projects/msalmon/hms-nhs-the-nautical-health-service/classify/workflow/18109 You can also open it by setting state in the Classify page component (or the Project Home component.) Subject selection can similarly be triggered by URLs or by setting state within the project app.

Ideally, the Classifier doesn't know anything about the page that it's embedded in, but here we've got a control for a feature of the project app.

**Reply by srallen** [reply]
Could we pass down a method like `addToCollection` as prop to the classifier that is used to open that modal? The classifier doesn't need to know anything else about it.

**Reply by snblickhan** [reply]
FYI I think this is an old version of the design for this modal -- the text is updated [here](https://projects.invisionapp.com/d/main#/console/12924056/438011817/preview).

**Reply by eatyourgreens** [reply]
A callback is probably the cleanest solution. The accessible version would look something like:

```js
function onClick(event) {
  const isCmdClick = event.metaKey
  if (!isCmdClick) {
    event.preventDefault()
    props.onSelectSubject(event)
  }
}

<a onClick={onClick} href={`/projects/${project.slug}/classify/workflow/${workflow.id}/subject-set/${subjectSet.id}`} >
  Subject Selection Modal
</a>
```

The link would be accessible in any assistive tech (screen readers, switches, Dragon etc.) The downside is that the subject set page URL is embedded in the classifier.

**Reply by eatyourgreens** [reply]
That link could also be a button, and still be accessible. We should be careful to distinguish between links and buttons in the visual design. At the moment, it looks like a link (to me.)

**Reply by srallen** [reply]
Could the callback just call a function in the project app that would open the modal and keep the project app responsible for updating the URL if something new is selected? I don't think this should directly be a link because someone may change their mind and want to cancel.

**Reply by eatyourgreens** [reply]
That's a good point. The workflow/subject set selection modal is deliberately designed so that it can't be closed once opened.
https://github.com/zooniverse/front-end-monorepo/pull/1976#pullrequestreview-569407971

I guess if it was a link, then you could use the Back button to cancel the action but I'm not sure about controlling modals with URLs either. The login and register forms open via URLs, which means they can be opened even if you're logged in.
https://www.zooniverse.org/projects/nora-dot-eisner/planet-hunters-tess/classify/workflow/11235?login=true
https://www.zooniverse.org/projects/nora-dot-eisner/planet-hunters-tess/classify/workflow/11235?register=true

**Reply by srallen** [reply]
That modal design makes sense if the only way to reach those is upon first page load, but now they're not, so they need to be cancelable I think now.

**Comment by eatyourgreens** [comment]
Now that https://github.com/zooniverse/front-end-monorepo/pull/2132 has merged, you can try out subject selection at https://fe-project.preview.zooniverse.org/projects/bogden/scarlets-and-blues/classify/workflow/17077 The Minutes workflow has subject sets that are indexed in Datasette.

The classifier hasn't been updated to use indexed selection yet, so it will crash after you select a subject.

**Reply by eatyourgreens** [reply]
Somewhere along the line, I've accidentally broken that URL. It's fixed in https://github.com/zooniverse/front-end-monorepo/pull/2169.

**Comment by eatyourgreens** [comment]
If we pass in the router as an external dependency, then accessible routing might look something like this (without having a direct dependency on `next/link`:

```js
/*
The router could also be injected into the store, like the API client is.
*/
const { router } = props
function onClick(event) {
  if (!event.metaKey) {
    event.preventDefault()
    const { href } = event.target
    router.push(href)
  }
}

<a href="/some/url">Client-side link in the parent app</a>
```

Marcy Sutton did a bunch of [user-testing of accessible client-side routing](https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/), for Gatsby, in 2019. That's worth reading because any solution we adopt will have to work with assistive technology. Gatsby published [details of their React implementation](https://www.gatsbyjs.com/blog/2020-02-10-accessible-client-side-routing-improvements/) in February last year.

**Reply by srallen** [reply]
Wouldn't this still couple the classifier to the routes and router of the project app?

**Reply by eatyourgreens** [reply]
Yeah, but I can’t think of another solution that’s accessible to assistive technology.

**Reply by srallen** [reply]
Is manual announcement not an option?

**Reply by srallen** [reply]
I still don't think it should be a direct link because of the need for cancellation as well.

**Reply by eatyourgreens** [reply]
If you’re navigating to a new URL you’ll need to either use a link, or emulate a link I’m afraid.

**Reply by eatyourgreens** [reply]
This is for the Project Home link, by the way.

**Reply by srallen** [reply]
This is not viable? https://github.com/zooniverse/front-end-monorepo/discussions/2165#discussioncomment-724724

**Reply by eatyourgreens** [reply]
Super useful description here of how page navigation works in screen readers, from Léonie Watson. https://github.com/gatsbyjs/gatsby/issues/5581#issuecomment-392919628

**Reply by srallen** [reply]
Taking a step back, why are we directing users back to the home page to begin with? Workflow selection can be done on the classify page.

**Reply by eatyourgreens** [reply]
Changing `window.location` would work if triggered from clicking a link.

**Reply by srallen** [reply]
Yeah, it's hard to keep track of what you're taking about here. Anchor link for project home page is fine with props to instruct how to build the link (though my question stands about why are we linking to the home page). Opening a modal is something I'm saying to do by callback.

**Reply by eatyourgreens** [reply]
>Taking a step back, why are we directing users back to the home page to begin with? Workflow selection can be done on the classify page.

That's a really good question. My ideal solution to handling links inside the classifier would be: don't have links to other pages inside the classifier. I'm not sure this dialog box needs a link to the home page either. All the other actions that it initiates could be handled by buttons, which change either the classifier state or the state of the parent page. Changing the state of the parent page could easily be done with either a callback from the classifier or the classifier sending a notification (or event) to its container.

**Reply by srallen** [reply]
@snblickhan do you have any insight why the design for the modal for when a subject has been retired has a link to the home page? If I had to guess, it would be because Becky was thinking that this is where workflow selection happens, but we now support this happening on the classify page directly, so navigating back to the home page is not necessary. Selecting a new workflow from on the classify page context has been a long requested feature, especially for Gravity Spy just for context, so staying on the page for this interaction I think is generally preferable. If there's another reason to go to the home page, that would be helpful for this discussion.

**Comment by srallen** [comment]
Having a decision about this soon would be beneficial to me as well. I'm getting closer to the point that I need to wire up #2143. There are two ways I'm thinking of doing this:

1. Accepting the new workflow would call a callback passed in as a prop to the classifier. The callback supplied by the project app would navigate to the newly assigned workflow id from the callback parameter. The classifier doesn't need to know anything about the routes or router.
2. Move the modal and UPP store to the project app and supply the classifier with callbacks as props to update the UPP as needed for keeping track of the last seen state for tutorials/mini-courses. Prompting and routing is handled by the project app entirely, but this makes the classifier incapable of doing workflow assignment independently. 

Users will always be able to cancel the prompt and have no routing or load of new workflow take place.

**Reply by eatyourgreens** [reply]
I think you'd get a clean separation of concerns with a system where the classifier handles the business logic of workflow assignment, but has no opinion about the mechanism by which a volunteer chooses a new workflow. When a new workflow becomes available, the classifier could notify its container through a callback, or by sending a notification message, with a payload containing the new workflow ID. In the project app, the Classify page could respond to that notification by showing a link to the new workflow in a popup (this is more or less how workflow selection works for Engaging Crowds.) Volunteers can either click on the workflow link, to go to the new workflow page, or cancel the popup, to continue as they were. When the workflow ID in the window location changes, client-side, the classifier's `workflowID` prop changes, triggering `selectWorkflow` here:
https://github.com/zooniverse/front-end-monorepo/blob/03b0c9a6dac4607aadf0e74facc689bac9fbbdf1/packages/lib-classifier/src/components/Classifier/Classifier.js#L114-L118

A second option would be to have the popup in the classifier itself, with the classifier notifying the project after workflow state updates in the classifier store. The window location could then update to reflect the new classifier state. In that case, the classifier is now deciding how you choose a new workflow (via a dialog box), which makes it less reusable as a component in my opinion. In that case, I think it would be difficult to communicate the change of window location to anyone using assistive technology (see the discussion of [accessible client-side routing](https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/) that Gatsby tested with people who had a variety of impairments.)

**Reply by eatyourgreens** [reply]
The project app loads UPP when you log in. It's been at the back of my mind to add that to the project store, as `user.projectPreferences`. I'm not sure if the browser caches that response, so that the cache is used when the classifier makes the same request. I have a feeling that auth'ed requests are never cached.
https://github.com/zooniverse/front-end-monorepo/blob/03b0c9a6dac4607aadf0e74facc689bac9fbbdf1/packages/app-project/stores/YourStats.js#L111-L112

`authorization` could be a property of the `User` model too.

**Reply by srallen** [reply]
Yes, I understand what I think the options are. What I was missing was that the project app is already loading up the UPP. I had forgotten that. I think we're at the point that a summary of the discussion here can be written up?

**Comment by srallen** [comment]
Ok, here's a summarization of where I think we are in this discussion Let me know if this is agreeable to everyone.

# Prompting a volunteer to work on something else

The context is we have persistently had problems with volunteers continuing to work on workflows where classifications are no longer needed. We have an improved design of the banners that appear with subjects, however, there is still a need to more specifically prompt to do something else.

## Possible prompt triggers

### All projects

The project or workflow is entirely finished.

### Projects using ordered subject set selection (like Engaging Crowds)

For humanities projects that follow the Engaging Crowds model subjects are ordered and can be loaded in the classifier regardless of retirement or already seen status. A subject can be loaded by way of a selection process provided in modal form on page load or by direct URL link.

### Workflow assignment (like Gravity Spy)

There are projects who wish to assign workflows. Some wish to do it like Gravity Spy where user weighing is analyzed by previously submitted classification quality and access to more complex workflows are blocked by this analysis. Others may have some other determining mechanism. The assigned workflow is received by a workflow id set in the user project preferences. 

If a volunteer gets directly linked to a workflow they do not have permission to access yet, then this will prompt the workflow selection modal to open instead with a message saying it's not available to them yet.

## Proposed solutions

A variety of scenarios exist where different prompts should load:

- if the project is finished, invite to work on another project. Simple version would be to link to the projects listing and a later iteration could suggest similar projects. The display of the modal and routing is handled entirely by the project app.
- if the workflow is finished, invite to work on another workflow if other active, unfinished workflows are available. This is simply a check on the finished status of the current loaded workflow and a load of the workflow selection modal with an added message that the currently chosen workflow is finished. This would be handled by the project app.
- if the user has been assigned a new workflow, then display a prompt inviting them to the new workflow
  - We should consider an option to not show this again for the session.
  - This is dependent on the on the UPP 
- if the subject is finished or already seen, which we won't know until the classifier loads, then call a callback which opens a modal messaging the subject status and the option to select a new subject. This could possibly load the subject set selection modal or the subject selection modal depending on the configuration of the project's workflow. This possibly may load where the user last left on if we have that information for the session.

**All prompts should be cancelable to allow the project, workflow, and subject to continue to load for education, demonstration, and citation purposes.** We might consider a future session demo and/or education mode for the classifier that toggles these various modals off for the session.

### General project application

This applies to all projects where the project or workflow may be finished. This check pauses the classifier from fully loading and waits for the volunteer selection on what to do?

### Ordered subject set application

There is a provided design for prompting when the subject is retired and already seen, but needs several modifications:

- The prompt should be cancelable as mentioned in the above section. This means a cancel button both in the header and in the modal content
- The prompt should not include next/previous button as we've decided this shouldn't be part of the task area because of potential confusion with the task navigation. Subject browsing is part of the discussion for #2166 and is a separate feature.
- The prompt should not include a link to the project home page. It's unclear what this is for and the selection modal for workflow, subject set, or subject selection can be opened on the classify page itself by way of callback from the classifier to the project app.

### Workflow selection application

Currently the user project preferences is being loaded twice. Once by the project app and once by the classifier. This is unnecessary. We should refactor these so that it is only loaded by the project and a callback is passed by prop to the classifier to make updates to the UPP as needed for tracking seen states for tutorials and mini-courses. The workflow assignment modal prompt will move to the project app and handle routing. 

The triggering of this prompt is still TBD and we may implement web sockets or a polling mechanism. There is already a callback for the project app to keep track of classification counts for the session so if we use polling, this can be easily implemented. For websockets, we need to add this anyway at some point to wire up talk notifications for the header in the project app, so could leverage this if Sugar is updated for this functionality.

**Reply by snblickhan** [reply]
This is great, but the only thing I’d add would be an option for if a subject set is finished — I think you’ve addressed this partially w/r/t ordered subject set application, but it’s more in the context of when a subject is done, not a whole set. The (admittedly rare) use case I’m thinking of is where someone is working on an almost-complete set, does a few classifications and then the set is done (or at least they’ve classified everything in that set) and need prompting to choose a new subject set (or workflow).

Regarding the link to the home page:

> If I had to guess, it would be because Becky was thinking that this is where workflow selection happens, but we now support this happening on the classify page directly, so navigating back to the home page is not necessary.

I think this was indeed the thought behind this, but I agree that if we can let them make this selection without leaving the classifier (regardless of 'level', i.e. subject/subject set/workflow) that would be ideal.

**Reply by eatyourgreens** [reply]
>The prompt should not include next/previous button as we've decided this shouldn't be part of the task area because of potential confusion with the task navigation. Subject browsing is part of the discussion for #2166 and is a separate feature.

I thought we'd agreed to implement the Next and Previous buttons in https://github.com/zooniverse/front-end-monorepo/issues/2027, rather than wait on new designs?

**Reply by eatyourgreens** [reply]
>if the subject is finished or already seen, which we won't know until the classifier loads,

For projects with indexed subject sets, we'll know the subject's classification status before the classifier loads. When you pick a subject to work on, the classification status is displayed alongside the link to the subject. See https://github.com/zooniverse/front-end-monorepo/issues/1955.

**Reply by eatyourgreens** [reply]
@snblickhan That's a good point about completed subject sets. At the moment, we can tell when a set is retired. In fact, I think we don't list retired sets because Cellect only sends back sets that are incomplete. For incomplete sets, I don't think we have a good way of detecting that a volunteer has completed that set (but not retired it), short of paging through the entire set and checking the user-seen status for each individual page.

**Reply by srallen** [reply]
If the subject is finished, do we want to assume they want to continue in the same workflow if there's more than one available unfinished? Depending on availability, we may want specific options to be displayed, i.e. "Select a new workflow", "Select a new subject set", etc. So I don't see a whole lot of difference in the UI presentation that we would have if it's just the subject that is finished vs the subject set, the variable will be in what is still available.

**Reply by srallen** [reply]
> I thought we'd agreed to implement the Next and Previous buttons in #2027, rather than wait on new designs?

@snblickhan can confirm, but my understanding is that we agreed it's a separate feature and would be implemented somewhere else around the classifier and specifically not in the task area. Getting a new subject is something that is independent from the retirement or already seen status, so shouldn't be coupled to this.

**Reply by snblickhan** [reply]
To clarify:
- Yes, the idea is that the 'browsing' arrows (forward/back button) is a separate feature, and needs to be accessible in cases which won't trigger the modal to open. Therefore, it shouldn't be on the modal.
- Additionally, the modal context is "This subject is done, take me to one I can work on" which is a separate need than "Take me to the next subject in sequence" -- that button should take volunteers to the next classifiable subject.
